### ISSUE TITLE
LONG-TERM: Fine-Tuned Small Language Model (SLM) Strategy for Extraction (~85% Cost Reduction)

### ISSUE THREAD

**Issue by BdM-15** [BdM-15/govcon-capture-vibe#19]
# LONG-TERM: Fine-Tuned Small Language Model (SLM) Strategy for Extraction

## Problem Statement

Current cloud LLM extraction costs **~$4 per RFP** using Grok-4-fast-reasoning. At scale, this becomes significant:
- 100 RFPs/year: $400
- 500 RFPs/year: $2,000
- 1,000 RFPs/year: $4,000

A fine-tuned local Small Language Model (SLM) could reduce extraction costs by **85-90%** while maintaining quality for the domain-specific entity extraction task.

### Why Fine-Tuning Makes Sense for This Use Case

1. **Highly Structured Task**: Entity extraction is pattern-based, not open-ended reasoning
2. **Domain-Specific**: Government contracting ontology is well-defined (18 entity types)
3. **High-Quality Training Data Available**: Existing extractions provide labeled pairs
4. **Repetitive Patterns**: RFPs follow standard structures (FAR, UCF, evaluation factors)
5. **Local Processing Preferred**: Sensitive proposal data shouldn't go to cloud

### Current vs Target Cost Analysis

| Metric | Cloud (Grok-4) | Local SLM | Savings |
|--------|----------------|-----------|---------|
| Per-chunk cost | $0.043 | ~$0.005 | 88% |
| Per-RFP cost | $4.00 | ~$0.60 | 85% |
| Annual (100 RFPs) | $400 | ~$60 | **$340** |
| Annual (500 RFPs) | $2,000 | ~$300 | **$1,700** |

### Two-Phase Approach

**Phase 1: Hybrid Architecture**
- Use fine-tuned SLM for extraction (high-volume, structured)
- Keep cloud LLM for query/reasoning (complex, lower-volume)

**Phase 2: Per-Entity Specialists**
- Train specialized models for different entity types
- Some entities may work with 3B models, others need 7B

### Training Data Generation Strategy

**Source: Existing High-Quality Extractions**

We have hundreds of validated extractions from production runs. Each extraction provides:
- Input: Document chunk text
- Output: Structured JSON with entities

```python
# Training data format
{
    "input": "The contractor shall provide 24/7 help desk support for 500 users with 4-hour response time for Priority 1 incidents. This service shall comply with ITIL v4 standards and maintain 99.9% uptime.",
    
    "output": {
        "entities": [
            {
                "entity_name": "24/7 Help Desk Support",
                "entity_type": "requirement",
                "criticality": "MANDATORY",
                "modal_verb": "shall",
                "labor_drivers": ["24/7 coverage", "500 users", "4-hour response"],
                "source_text": "The contractor shall provide 24/7 help desk support..."
            },
            {
                "entity_name": "99.9% Uptime Standard",
                "entity_type": "performance_metric",
                "threshold": "99.9%",
                "measurement_method": "Uptime monitoring",
                "source_text": "...maintain 99.9% uptime"
            }
        ]
    }
}
```

### Candidate Models

| Model | Size | VRAM | Inference Speed | Use Case |
|-------|------|------|-----------------|----------|
| Phi-3 Mini | 3.8B | 4GB | Very fast | Simple entities (clause, section) |
| Llama 3.2 | 3B | 4GB | Very fast | Basic extraction |
| Mistral 7B | 7B | 8GB | Fast | Complex entities (requirement, evaluation_factor) |
| Llama 3.2 | 8B | 10GB | Moderate | Full extraction |
| Qwen 2.5 | 14B | 16GB | Moderate | High accuracy extraction |

### Per-Entity Specialist Strategy

Different entity types have different complexity:

**Tier 1: Simple Patterns (3B model)**
- `section` - Document structure markers
- `clause` - FAR/DFARS citations
- `organization` - Named entities

**Tier 2: Moderate Complexity (7B model)**
- `requirement` - Modal verbs, criticality
- `deliverable` - CDRL patterns
- `concept` - Domain terminology

**Tier 3: Complex Inference (7B-14B model)**
- `evaluation_factor` - Weights, subfactors, hierarchy
- `submission_instruction` - Format requirements, page limits
- `performance_metric` - Thresholds, measurement methods
- `strategic_theme` - Customer hot buttons, discriminators

### Phase 1: Training Data Export (Week 1-2)

```python
# tools/export_training_data.py
async def export_training_pairs(neo4j_io, output_dir: str):
    """Export extraction pairs from Neo4j for fine-tuning."""
    
    # Get all entities with their source chunks
    entities = neo4j_io.get_all_entities_with_source()
    
    training_pairs = []
    for entity in entities:
        pair = {
            "input": entity["source_chunk_text"],
            "output": {
                "entities": [entity_to_dict(entity)]
            }
        }
        training_pairs.append(pair)
    
    # Save in fine-tuning format (JSONL)
    with open(f"{output_dir}/training_data.jsonl", "w") as f:
        for pair in training_pairs:
            f.write(json.dumps(pair) + "\n")
```

**Target: 1,000+ training pairs** from existing extractions

### Phase 2: Model Selection and Fine-Tuning (Week 3-4)

**Fine-tuning approach: QLoRA (Quantized Low-Rank Adaptation)**
- 4-bit quantization for memory efficiency
- LoRA adapters for targeted fine-tuning
- RTX 4060 8GB can handle 7B model fine-tuning

```bash
# Example fine-tuning command (using Unsloth)
python finetune.py \
    --base_model="mistralai/Mistral-7B-v0.3" \
    --dataset="training_data.jsonl" \
    --output_dir="models/govcon_extractor_7b" \
    --num_epochs=3 \
    --lora_r=16 \
    --lora_alpha=32 \
    --batch_size=4
```

### Phase 3: Evaluation Framework (Week 5)

```python
# tools/evaluate_slm.py
async def evaluate_extraction_quality(slm_extractor, cloud_extractor, test_chunks):
    """Compare SLM vs cloud extraction quality."""
    
    metrics = {
        "entity_count_parity": [],
        "entity_type_accuracy": [],
        "field_completeness": [],
        "relationship_accuracy": []
    }
    
    for chunk in test_chunks:
        slm_result = await slm_extractor.extract(chunk)
        cloud_result = await cloud_extractor.extract(chunk)
        
        # Compare results
        metrics["entity_count_parity"].append(
            len(slm_result.entities) / len(cloud_result.entities)
        )
        # ... additional metrics
    
    return aggregate_metrics(metrics)
```

**Acceptance threshold**: ≥95% parity with cloud extraction

### Phase 4: Hybrid Integration (Week 6-8)

```python
# src/extraction/hybrid_extractor.py
class HybridExtractor:
    def __init__(self):
        self.slm_extractor = SLMExtractor("models/govcon_extractor_7b")
        self.cloud_extractor = CloudExtractor()  # Grok-4 fallback
    
    async def extract(self, chunk: str, entity_types: List[str] = None) -> ExtractionResult:
        # Use SLM for supported entity types
        if self._can_use_slm(entity_types):
            result = await self.slm_extractor.extract(chunk)
            
            # Validate result quality
            if self._passes_quality_check(result):
                return result
            else:
                logger.warning("SLM extraction below threshold, falling back to cloud")
        
        # Fallback to cloud for complex/unsupported types
        return await self.cloud_extractor.extract(chunk)
```

### Phase 5: Per-Entity Specialists (Week 9-12)

Train specialized models for different entity groups:

```python
# src/extraction/specialist_extractor.py
class SpecialistExtractor:
    def __init__(self):
        self.models = {
            "simple": load_model("models/govcon_simple_3b"),    # section, clause
            "moderate": load_model("models/govcon_moderate_7b"), # requirement, deliverable
            "complex": load_model("models/govcon_complex_14b")   # eval_factor, strategic_theme
        }
        
        self.entity_routing = {
            "section": "simple",
            "clause": "simple",
            "organization": "simple",
            "requirement": "moderate",
            "deliverable": "moderate",
            "evaluation_factor": "complex",
            "submission_instruction": "complex",
            "performance_metric": "complex",
            "strategic_theme": "complex"
        }
    
    async def extract(self, chunk: str) -> ExtractionResult:
        # Run all specialists in parallel
        results = await asyncio.gather(
            self.models["simple"].extract(chunk),
            self.models["moderate"].extract(chunk),
            self.models["complex"].extract(chunk)
        )
        
        # Merge results
        return merge_extraction_results(results)
```

## Hardware Requirements

**Current System:**
- GPU: RTX 4060 8GB VRAM
- RAM: 64GB DDR5
- CPU: i9-14900HX

**Capability Assessment:**

| Task | VRAM Required | Current VRAM | Feasible? |
|------|---------------|--------------|-----------|
| 7B inference (4-bit) | 4GB | 8GB | ✅ Yes |
| 7B fine-tuning (QLoRA) | 6GB | 8GB | ✅ Yes |
| 14B inference (4-bit) | 8GB | 8GB | ⚠️ Tight |
| 14B fine-tuning (QLoRA) | 12GB | 8GB | ❌ Need offload |

**Recommendation**: Start with 7B models, consider 14B only if 7B quality insufficient

### Quality Metrics

| Metric | Target | Measurement |
|--------|--------|-------------|
| Entity count parity | ≥95% | SLM entities / Cloud entities |
| Entity type accuracy | ≥98% | Correct type / Total entities |
| Field completeness | ≥90% | Non-null fields / Expected fields |
| Relationship quality | ≥90% | Valid relationships / Total relationships |

### Cost Metrics

| Metric | Target | Measurement |
|--------|--------|-------------|
| Per-RFP cost | ≤$1.00 | Compute cost + cloud fallback |
| Cloud fallback rate | ≤10% | Cloud calls / Total calls |
| ROI breakeven | ≤50 RFPs | Development cost / Per-RFP savings |

### Performance Metrics

| Metric | Target | Measurement |
|--------|--------|-------------|
| Extraction latency | ≤cloud | Time per chunk |
| Total RFP time | ≤60 min | End-to-end processing |
| GPU utilization | ≤80% | During extraction |

## Acceptance Criteria

- [ ] 1,000+ training pairs exported from production extractions
- [ ] Fine-tuned 7B model achieves ≥95% entity parity with cloud
- [ ] Hybrid extractor implemented with automatic fallback
- [ ] Per-RFP cost reduced by ≥75%
- [ ] Query/reasoning remains on cloud LLM
- [ ] Per-entity specialists evaluated for ROI
- [ ] Documentation for model training and deployment

## Files to Create/Modify

| File | Purpose |
|------|---------|
| `tools/export_training_data.py` | Export training pairs from Neo4j |
| `tools/finetune_slm.py` | Fine-tuning script |
| `tools/evaluate_slm.py` | Quality evaluation framework |
| `src/extraction/slm_extractor.py` | Local SLM inference |
| `src/extraction/hybrid_extractor.py` | Cloud + SLM hybrid |
| `models/` | Fine-tuned model storage |

## Timeline

| Week | Milestone |
|------|-----------|
| 1-2 | Training data export, data quality validation |
| 3-4 | Fine-tune 7B model, initial evaluation |
| 5 | Quality evaluation, identify gaps |
| 6-8 | Hybrid integration, fallback logic |
| 9-12 | Per-entity specialists (if ROI positive) |

## Risks and Mitigations

| Risk | Likelihood | Impact | Mitigation |
|------|------------|--------|------------|
| Quality below target | Medium | High | Cloud fallback, iterative training |
| VRAM insufficient | Low | Medium | Use smaller models, quantization |
| Training data bias | Medium | Medium | Diverse RFP sources, validation set |
| Maintenance overhead | Medium | Low | Automated retraining pipeline |

## Priority

**LONG-TERM** - Highest potential cost savings, requires significant development

- Estimated development: 8-12 weeks
- ROI breakeven: ~50 RFPs (with $3.40 savings per RFP)
- Strategic value: Local processing for sensitive data

## Related Issues

- Issue #14: Prompt compression (reduces SLM prompt complexity)
- Issue #17: Parallel processing (SLM enables local parallelism)
- Issue #6: Pydantic schema (structured output for SLM training)
- Issue #9: Entity extraction prompts (training data quality)

**Comment by BdM-15** [comment]
## Related Documentation & Strategic Benefits Connection

This issue is part of the **Training Data & Fine-Tuning Foundation** strategic benefit from the README.md roadmap.

### Comprehensive Documentation Available

📄 **`docs/future_features/performance/FINE_TUNING_ROADMAP.md`** (400+ lines)

This document provides extensive additional detail on:

1. **Training Data Collection Strategy** (Phase 1)
   - Entity-action-context extraction patterns
   - Input/output pair formatting
   - Source chunk to entity mapping
   - Target: 500-1000 initial examples

2. **Model Candidates with Evaluation Criteria**
   - Qwen 2.5 7B, Mistral 7B, Llama 3.1 8B
   - Benchmark considerations
   - Quantization strategies (INT8, INT4)

3. **Error Pattern Analysis Framework**
   - Categorization: extraction errors, classification errors, relationship errors
   - Severity weighting for training prioritization
   - Iterative improvement cycles

4. **Phase 2 Status**
   - Currently DEFERRED in favor of cloud acceleration
   - Will revisit after processing 50+ RFPs with validated extractions
   - Focus on prompt optimization (#14) first

### Connected Issues

| Issue | Connection |
|-------|------------|
| #14 | Prompt Compression - reduces SLM prompt complexity |
| #22 | Performance Analytics - training data feedback loop |
| #9 | Entity Extraction Prompts - training data quality |
| #6 | Pydantic Schema - structured output format |

### User Feedback Loop for Training Data

Issue #22 (Performance Analytics Dashboard) includes the feedback collection mechanism that feeds into this training pipeline:

```python
@app.post("/feedback")
async def submit_feedback(
    query_id: str,
    rating: Literal["good", "bad", "partial"],
    correction: Optional[str] = None
):
    """Collect user feedback for training data pipeline."""
    await training_data_collector.log_feedback(...)
```

This creates the continuous improvement cycle:
1. **Extraction** → User reviews → **Correction** → Training data
2. **Training data** → Fine-tuning → **Improved model** → Better extraction
